SCons/Scanner/__init__.py

```diff
@@ -201,23 +213,24 @@ def __call__(self, node, env, path=()):
         if self.scan_check and not self.scan_check(node, env):
             return []
 
-        self = self.select(node)
+        # here we may morph into a different Scanner instance:
+        self = self.select(node)  # pylint: disable=self-cls-assignment
 
         if self.argument is not _null:
             node_list = self.function(node, env, path, self.argument)
         else:
             node_list = self.function(node, env, path)
 
-        kw = {}
-        if hasattr(node, 'dir'):
-            kw['directory'] = node.dir
-        node_factory = env.get_factory(self.node_factory)
-        nodes = []
-        for l in node_list:
-            if self.node_class and not isinstance(l, self.node_class):
-                l = node_factory(l, **kw)
-            nodes.append(l)
-        return nodes
+        if self.node_class:
```

Comment:
Can you add some comments here what the code does?

Reply:
the *change* rearranges so we don't have to do two checks for each node, and thus can move the list productions into a comprehension.

Reply:
Oh I meant comments in the code.

Reply:
Hmmm, as I look at this I'm not sure my refactoring is correct. Let me think about it, we can leave this for the next release.